addons/sale_coupon/models/sale_order.py

```diff
@@ -185,8 +246,14 @@ def _get_reward_values_discount(self, program):
 
             # when processing lines we should not discount more than the order remaining total
             currently_discounted_amount = 0
+
+            tax_groups = set([line.tax_id for line in lines])
+            max_discount_per_tax_groups = {tax_ids: self._get_max_reward_values_per_tax(program, tax_ids) for tax_ids in tax_groups}
+
             for line in lines:
                 discount_line_amount = min(self._get_reward_values_discount_percentage_per_line(program, line), amount_total - currently_discounted_amount)
+                discount_line_amount = min(max_discount_per_tax_groups[line.tax_id], discount_line_amount)
+                max_discount_per_tax_groups[line.tax_id] -= discount_line_amount if discount_line_amount > 0 else 0
```

Comment:
Why do you need to alter the % discount mechanism?
How can a discount_line_amount (which will always be < than the SOL amount as it is a %age of it) be greater than the amount of multiple SOL including itself?
Is that because of the fixed discount amount?

Reply:
In case you apply a fixed amount before a % amount there is an issue without this check.
In most of the code, there is no issue as the fix force the application of % discount before fixed one but currently without this, if you applied a fixed price discount on a product and the a % which would make the price go lower than 0, you could have a negative total price which this line prevent.

After further thought, we could force the recomputation of discount line for specific product too and get rid of this.